File: Списки/Spiski.cpp
// шаблон класса ListNode
template<class T> class ListNode : public Node {
	public:
    T _val;
	ListNode(T val);
	friend class List<T>;
};

// конструктор ListNode
template<class T> ListNode::ListNode(T val) :
	_val(val)
{
}

// шаблон класса List
template<class T> class List{
	private:
		ListNode<T> *header;
		ListNode<T> *win;
		int _lenght;
	public:
		List (void);
		~List(void);
		T insert(T);
		T append(T);
		List *append(List *);
		T prepend(T);
		T remove(void);
		void val(T);
		T val (void);
		T next (void);
		T prev (void);
		T first (void);
		T last (void);
		int length (void);
		bool isFirst (void);
		bool isLast(void);
		bool isHead (void);
};

// конструктор создает и инициализирует пустой список, представленный одним  головным узлом, со ссылками на себя
template<class T> List<T>::List(void) :
	_length(0)
{
	header = new ListNode<T> (Null);
	win = header;
}

// разрушает узлы связанного списка (не разрушает сами элементы данных)
template<class T> List<T>::~List(void)
{
	while (length() > 0) {
		first();
		remove();
	}
	delete header;
}

// заносит новыый элемент после окна и возвращает указатель на новый элемент
template<class T> T List<T>::insert(T val)
{
	win ->insert(new ListNode<T>(val));
	++ _length;
	return val;
	
}

// prepend и append вставляют новый элемент в начало или конец списка 
// соответсвенно и возращают указатель на новый элемент
template<class T> T List<T>::prepend(T val)
{
	header -> insert(new ListNode<T> (val));
	++_length;
	return val;
}


template<class T> T List<T>::append(T val)
{
	header ->prev() -> insert (new ListNode<T> (val));
	++_length;
	return val;
}

// вторая версия  функции append - добавление списка 1 в конец  
// текущего списка, первый элемент списка 1 помещается после последнего
// элемента текущего списка. В процессе выполнения список 1 - пустой.
template<class T> List<T>* List<T>::append(List<T> *l)
{
	ListNode<T> *a = (ListNode<T>*) header -> prev();
	a -> splice (l->header);
	_length += 1; 1-> _length;
	l ->header->remove();
	l->_length = 0;
	l->win = header;
	return this;
}

// удаляет элемент в окне, перемещает окно в предыдущую позицию и возвращает 
// указатель на только что удаленный элемент.
// ПРИМЕЧАНИЕ! если окно находится  в головной позиции, то таких действий не производится
template<class T> T List<T>::remove(void)
{
	if (win == header)
		return NULL;
	void *val = win ->_val;
	win = (ListNode<T>*)win->prev();
	delete (listNode<T>*)win->next()->remove();
	--_length;
	return val;

}

// замена элемента, находящегося  в текущем окне, на элемент v
// елси в голове, то дейстие не производится
void List<T>::val (T v)
{
	if (win != header)
		win -> _val = v;
}

// если val без аргументов, то возвращется элемент в окне, либо NULL, если окно в головной позиции
template<class T> T List<T>::val(void)
{
	return win->_val;
}

// next и prev перемещают окно в следующую или предыдущую позиции
// каждая из них возвращает указатель на элемент, расположенный в новой позиции окна
template<class T> T List<T>::next(void)
{
	win = (ListNode<T>*) win->next();
	return win->_val;
}

template<class T> T List<T>::prev(void)
{
	win = (ListNode<T>*) win->prev();
	return win->_val;
}

// переносят окно в первую и последнюю позицию соотвественно
template<class T> T List<T>::first(void)
{
	win = (ListNode<T>*) header->next();
	return win->_val;
}

template<class T> T List<T>::last(void)
{
	win = (ListNode<T>*) header->prev();
	return win->_val;
}

// значение длины списка
template<class T> int List<T>::length(void)
{
	return _length;
}

// isFirst, isLast, isHead возвращают TRUE только в случае выполнения условий,
// что окно находится либо в первой, либо последней, либо головной позиции соответсвенно
template<class T> bool List<T>::isFirst(void)
{
	return ((win ==header ->next()) && (_legth > 0));
}


template<class T> bool List<T>::isLast(void)
{
	return ((win ==header ->()) prev&& (_legth > 0));
}


template<class T> bool List<T>::isHead(void)
{
	return (win == header);
}


// Примеры списков 

// функция загружает n элементов массива a в список и возвращает указатель на этот список
template<class T> List<T> *arrayToList(T a[], int n)
{
	List<T> *s = new List<T>;
	for (int i = 0; i < n; i++)
		s ->append(a[i]);
	return s;
}

// функция возвращает наименьший элемент в списке s
template<class T> T leastItem(List<T> &s, int(*cmp) (T,T))
{
	int i;
	if (s.length() == 0)
		return NULL;
	T v = s.first();
	for (s.next(); !s.isHead(); s.next())
		if (cmp(s.val(), v) < 0)
			v = s.val();
	return v;
}

File: Списки/Spiski.js
// шаблон класса ListNode
class ListNode {
    constructor(val) {
        this._val = val; // значение узла
        this._next = this; // связь к последующему узлу
        this._prev = this; // связь к предшествующему узлу
    }

    next() {
        return this._next; // перемещение к последующему узлу
    }

    prev() {
        return this._prev; // перемещение к предшествующему узлу
    }

    insert(node) {
        // включение узла в связанный список
        let c = this._next;
        node._next = c;
        node._prev = this;
        this._next = node;
        c._prev = node;
        return node;
    }

    remove() {
        // удаление узла из списка
        this._prev._next = this._next;
        this._next._prev = this._prev;
        this._next = this._prev = this;
        return this;
    }

    splice(node) {
        // слияние/разъединение двух узлов
        let an = this._next;
        let bn = node._next;
        this._next = bn;
        node._next = an;
        an._prev = node;
        bn._prev = this;
    }

    toString() {
        return `ListNode(${this._val})`; // для удобства вывода
    }
}

// шаблон класса List
class List {
    constructor() {
        // конструктор создает и инициализирует пустой список, представленный одним головным узлом, со ссылками на себя
        this._length = 0;
        this.header = new ListNode(null);
        this.win = this.header;
    }

    // заносит новый элемент после окна и возвращает указатель на новый элемент
    insert(val) {
        let newNode = new ListNode(val);
        this.win.insert(newNode);
        this._length++;
        return val;
    }

    // prepend вставляет новый элемент в начало списка и возвращает указатель на новый элемент
    prepend(val) {
        let newNode = new ListNode(val);
        this.header.insert(newNode);
        this._length++;
        return val;
    }

    // append вставляет новый элемент в конец списка и возвращает указатель на новый элемент
    append(val) {
        let newNode = new ListNode(val);
        this.header.prev().insert(newNode);
        this._length++;
        return val;
    }

    // добавление списка l в конец текущего списка
    appendList(list) {
        let a = this.header.prev();
        a.splice(list.header);
        this._length += list._length;
        list.header.remove();
        list._length = 0;
        list.win = list.header;
        return this;
    }

    // удаляет элемент в окне, перемещает окно в предыдущую позицию и возвращает указатель на только что удаленный элемент
    remove() {
        if (this.win === this.header) {
            return null;
        }
        let val = this.win._val;
        this.win = this.win.prev();
        this.win.next().remove();
        this._length--;
        return val;
    }

    // замена элемента, находящегося в текущем окне, на элемент v
    setVal(val) {
        if (this.win !== this.header) {
            this.win._val = val;
        }
    }

    // возвращает элемент в окне, либо null, если окно в головной позиции
    getVal() {
        return this.win._val;
    }

    // перемещает окно в следующую позицию и возвращает указатель на элемент в новой позиции окна
    next() {
        this.win = this.win.next();
        return this.win._val;
    }

    // перемещает окно в предыдущую позицию и возвращает указатель на элемент в новой позиции окна
    prev() {
        this.win = this.win.prev();
        return this.win._val;
    }

    // переносит окно в первую позицию и возвращает указатель на элемент в новой позиции окна
    first() {
        this.win = this.header.next();
        return this.win._val;
    }

    // переносит окно в последнюю позицию и возвращает указатель на элемент в новой позиции окна
    last() {
        this.win = this.header.prev();
        return this.win._val;
    }

    // возвращает длину списка
    length() {
        return this._length;
    }

    // возвращает true, если окно находится в первой позиции
    isFirst() {
        return (this.win === this.header.next() && this._length > 0);
    }

    // возвращает true, если окно находится в последней позиции
    isLast() {
        return (this.win === this.header.prev() && this._length > 0);
    }

    // возвращает true, если окно находится в головной позиции
    isHead() {
        return (this.win === this.header);
    }
}

// функция загружает n элементов массива в список и возвращает указатель на этот список
function arrayToList(arr) {
    let list = new List();
    for (let val of arr) {
        list.append(val);
    }
    return list;
}

// функция возвращает наименьший элемент в списке s
function leastItem(list, cmp) {
    if (list.length() === 0) {
        return null;
    }
    let v = list.first();
    for (list.next(); !list.isHead(); list.next()) {
        if (cmp(list.getVal(), v) < 0) {
            v = list.getVal();
        }
    }
    return v;
}

const main = () => {
    let list = new List();
    list.append(10);
    list.append(20);
    list.prepend(5);

    console.log("First element:", list.first()); // 5
    console.log("Last element:", list.last()); // 20

    let arr = [3, 1, 4, 1, 5];
    let newList = arrayToList(arr);
    console.log("First element of new list:", newList.first()); // 3

    let least = leastItem(newList, (a, b) => a - b);
    console.log("Least item in the new list:", least); // 1
};

main();
